low_level_platform/impl/src/lf_STM32f4_support.c

```diff
@@ -0,0 +1,238 @@
+  //      Which means period of 1 microsecond
+  TIM5->PSC = 15;
+
+  // Setup ISR to increment upper bits
+  TIM5->DIER |= TIM_DIER_CC1IE;
+  NVIC_EnableIRQ(TIM5_IRQn);
+
+  /* This is to make the Prescaler actually work
+   *  For some reason, the Prescaler only kicks in once the timer has reset once.
+   *  Thus, the current solution is to manually ret the value to a really large
+   *  and force it to reset once. Its really jank but its the only way I could
+   *  find to make it work
+   */
+  TIM5->CNT = 0xFFFFFFFE;
+}
+
+// ISR for handling timer overflow -> We increment the upper timer
+void TIM5_IRQHandler(void) {
+  if (TIM5->SR & (1 << 1)) {
+    TIM5->SR &= ~(1 << 1);
+    _lf_time_us_high += 1;
+  }
+}
+
+/**
+ * Write the time since boot into time variable
+ */
+int _lf_clock_gettime(instant_t* t) {
+  // Timer is cooked
+  if (!t) {
+    return 1;
+  }
+  // Get the current microseconds from TIM5
+  uint32_t _lf_time_us_low = TIM5->CNT;
+
+  // Combine upper and lower timers (Yoinked from lf_nrf52 support)
+  uint64_t now_us = COMBINE_HI_LO((_lf_time_us_high - 1), _lf_time_us_low);
+  *t = ((instant_t)now_us) * 1000;
+  return 0;
+}
+
+/**
+ * Make the STM32 sleep for set nanoseconds
+ * Based on the lf_nrf52 support implementation
+ */
+int lf_sleep(interval_t sleep_duration) {
+  instant_t target_time;
+  instant_t current_time;
+
+  _lf_clock_gettime(&current_time);
+  target_time = current_time + sleep_duration;
+  // HAL_Delay only supports miliseconds. We try to use that for as long as possible
+  //      before switching to another meothd for finer tuned delay times
+  long delaytime_ms = sleep_duration / 1000000;
+  HAL_Delay(delaytime_ms);
+
+  while (current_time <= target_time)
+    _lf_clock_gettime(&current_time);
+
+  return 0;
+}
+
+/**
+ * Make the STM32 sleep for set nanoseconds
+ * Based on the lf_nrf52 support implementation
+ */
+static void lf_busy_wait_until(instant_t wakeup_time) {
+  instant_t current_time;
+  _lf_clock_gettime(&current_time);
+
+  // We repurpose the lf_sleep function here, just to better streamline the code
+  interval_t sleep_duration = wakeup_time - current_time;
+  lf_sleep(sleep_duration);
+}
+
+// I am pretty sure this function doesnt work
+/*  sleep until wakeup time but wake up if there is an async event
+ */
+int _lf_interruptable_sleep_until_locked(environment_t* env, instant_t wakeup_time) {
+  // Get the current time and sleep time
+  instant_t now;
+  _lf_clock_gettime(&now);
+  interval_t duration = wakeup_time - now;
+
+  // Edge case handling for super small duration
+  if (duration <= 0) {
+    return 0;
+  } else if (duration < LF_MIN_SLEEP_NS) {
+    lf_busy_wait_until(wakeup_time);
+    return 0;
+  }
+
+  // Enable interrupts and prepare to wait
+  _lf_async_event = false;
+  lf_enable_interrupts_nested();
+
+  do {
+    _lf_clock_gettime(&now);
+    // Exit when the timer is up or there is an exception
+  } while (!_lf_async_event && (now < wakeup_time));
```

Comment:
This should be low-power sleep, not busy-waiting. Is there anything in the HAL for doing low-power sleep?

Reply:
There seems to be a low power sleep mode, but its only interrupt based. It seems most timed delay implementations just use a while loop. Im not sure it itll be worth it trying to create some workaround now? or just merge now and maybe improve the implementation later.